src/main/java/ch/njol/skript/expressions/ExprScript.java

```diff
@@ -18,76 +18,142 @@
  */
 package ch.njol.skript.expressions;
 
+import ch.njol.skript.ScriptLoader;
 import ch.njol.skript.Skript;
+import ch.njol.skript.SkriptCommand;
 import ch.njol.skript.doc.Description;
-import ch.njol.skript.doc.Events;
 import ch.njol.skript.doc.Examples;
 import ch.njol.skript.doc.Name;
 import ch.njol.skript.doc.Since;
 import ch.njol.skript.lang.Expression;
 import ch.njol.skript.lang.ExpressionType;
 import ch.njol.skript.lang.parser.ParserInstance;
+import org.jetbrains.annotations.Nullable;
 import org.skriptlang.skript.lang.script.Script;
 import ch.njol.skript.lang.SkriptParser.ParseResult;
 import ch.njol.skript.lang.util.SimpleExpression;
 import ch.njol.util.Kleenean;
 import org.bukkit.event.Event;
-import org.eclipse.jdt.annotation.Nullable;
 
-@Name("Script Name")
-@Description("Holds the current script's name (the file name without '.sk').")
+import java.io.File;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+@Name("Script")
+@Description({"The current script, or a script from its (file) name.",
+	"If the script is enabled or disabled (or reloaded) this reference will become invalid.",
+	"Therefore, it is recommended to obtain a script reference <em>when needed</em>."})
 @Examples({
+	"on script load:",
+	"\tbroadcast \"Loaded %the current script%\"",
 	"on script load:",
 	"\tset {running::%script%} to true",
 	"on script unload:",
-	"\tset {running::%script%} to false"
+	"\tset {running::%script%} to false",
+	"set {script} to the script named \"weather.sk\"",
+	"loop the scripts in directory \"quests/\":",
+	"\tenable loop-value"
 })
 @Since("2.0")
-@Events("Script Load/Unload")
-public class ExprScript extends SimpleExpression<String> {
-	
+public class ExprScript extends SimpleExpression<Script> {
+
 	static {
-		Skript.registerExpression(ExprScript.class, String.class, ExpressionType.SIMPLE,
-			"[the] script[['s] name]",
-			"name of [the] script"
+		Skript.registerExpression(ExprScript.class, Script.class, ExpressionType.SIMPLE,
+			"[the] [current] script",
+			"[the] script[s] [named] %strings%",
+			"[the] scripts in [directory|folder] %string%"
 		);
 	}
-	
-	@SuppressWarnings("NotNullFieldNotInitialized")
-	private String name;
-	
+
+	private @Nullable Script script;
+	private @Nullable Expression<String> name;
+	private int mode;
```

Comment:
Then I'd need an extra flag for the other mode, it was just simpler this way. I'll add a comment to say what it means though.

Reply:
For which mode? Since there are three patterns, we already know if it is the first one by whether `script` is null. `isDirectory` can differentiate between the two patterns where `name` is not null.

Reply:
Bumping this

Reply:
Given that `mode` is always compared against `2`, this should be easy enough.